core/src/main/java/com/graphhopper/storage/index/Snap.java

```diff
@@ -145,12 +146,28 @@ public void calcSnappedPoint(DistanceCalc distCalc) {
         double queryLat = getQueryPoint().lat, queryLon = getQueryPoint().lon;
         double adjLat = fullPL.getLat(wayIndex + 1), adjLon = fullPL.getLon(wayIndex + 1);
         if (distCalc.validEdgeDistance(queryLat, queryLon, tmpLat, tmpLon, adjLat, adjLon)) {
-            GHPoint tmpPoint = distCalc.calcCrossingPointToEdge(queryLat, queryLon, tmpLat, tmpLon, adjLat, adjLon);
+            GHPoint crossingPoint = distCalc.calcCrossingPointToEdge(queryLat, queryLon, tmpLat, tmpLon, adjLat, adjLon);
             double adjEle = fullPL.getEle(wayIndex + 1);
-            snappedPoint = new GHPoint3D(tmpPoint.lat, tmpPoint.lon, (tmpEle + adjEle) / 2);
-        } else
-            // outside of edge boundaries
-            snappedPoint = new GHPoint3D(tmpLat, tmpLon, tmpEle);
+
+            // We want to prevent extra virtual nodes and very short virtual edges in case the snap/crossing point is
+            // very close to a tower node. Since we delayed the calculation of the crossing point until here, we need
+            // to correct the Snap.Position in these cases. Note that it is possible that the query point is very far
+            // from the tower node, but the crossing point is still very close to it.
+            // if crossingPoint is too close to a pillar or tower node we have to fix the snappedPosition and wayIndex to make it consistent with GHPoint::equals
+            if (NumHelper.equalsEps(crossingPoint.lat, tmpLat) && NumHelper.equalsEps(crossingPoint.lon, tmpLon)) {
```

Comment:
Why does this have to be consistent, or is even related to `GHPoint#equals`? Also I would rather use `Math.abs() > threshold` here so we do not rely on whatever `equalsEps` does, which could be used for something unrelated to this code.

Reply:
Can't we use a `tmpPoint` and directly invoke `crossingPoint.equals(tmpPoint)` instead?

Reply:
I did not really get why this needs to be consistent, but since we now use a dedicated function (considerEqual) this works for me.

Reply:
This code in QueryOverlayBuilder:

```java
                    // no new virtual nodes if very close ("snap" together)
                    if (Snap.considerEqual(prevPoint.lat, prevPoint.lon, currSnapped.lat, currSnapped.lon)) {
                        res.setClosestNode(prevNodeId);
                        continue;
                    }
```

is supposed to make sure no extra virtual nodes are created when two snapping points are close to each other. An extreme case would be this:

```
    p
----x-----
    q
```                    
where the two points p and q are left/right of the edge and the snapping points are very close (or even equal). I still don't see why this should be related to the code in Snap.java that makes sure that edge snaps close to tower nodes are converted to tower snaps.

The reason the test fails once we use another tolerance here is that we do not update the snapping point  `res.getSnappedPoint()`: We simply update the closest node to be the one from the previous snap, but forget to update the snapped point (and the query distance). Basically, we first do the snapping, but then realize we want the snapping to be slightly different (close snaps should snap to the same virtual node), but do not update the modified snap correctly.

Reply:
You mean we should not do it in QueryOverlayBuilder? Why not? We cannot really do it before we got all the snaps and the crossing points, because first we need to know which query points should use the same virtual node?

Reply:
If we don't want to modify the snaps we need to distinguish between 'Snaps' (immutable result return types of the location index) and the things used by QueryGraph/Overlay to create virtual nodes/edges. But currently they are the same.

Reply:
Or if we want to use the same objects but don't want to modify the fields that were determined by location index, we need to add separate fields for this purpose (snappedPoint + resolvedPoint or something).

Reply:
Yes, in the current state I think this is the best we can do.